common/src/Microsoft.Azure.IIoT.AspNetCore/src/Auth/Clients/MsalUserTokenClient.cs

```diff
@@ -0,0 +1,305 @@
+    using Microsoft.AspNetCore.Authentication;
+    using Microsoft.Identity.Client;
+    using Serilog;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.IdentityModel.Tokens.Jwt;
+    using System.Security.Claims;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Authenticate on behalf of current logged in claims principal to another
+    /// service. This uses the behalf_of flow defined in xxx.
+    /// </summary>
+    public class MsalUserTokenClient : ITokenClient, IUserTokenClient {
+
+        /// <inheritdoc/>
+        public string Provider => AuthProvider.AzureAD;
+
+        /// <summary>
+        /// Create auth provider. Need to also inject the http context accessor
+        /// to be able to get at the http context here.
+        /// </summary>
+        /// <param name="ctx"></param>
+        /// <param name="cache"></param>
+        /// <param name="schemes"></param>
+        /// <param name="config"></param>
+        /// <param name="logger"></param>
+        public MsalUserTokenClient(IHttpContextAccessor ctx, ICache cache,
+            IAuthenticationSchemeProvider schemes, IClientAuthConfig config, ILogger logger) {
+            _schemes = schemes ?? throw new ArgumentNullException(nameof(schemes));
+            _ctx = ctx ?? throw new ArgumentNullException(nameof(ctx));
+            _cache = cache ?? throw new ArgumentNullException(nameof(cache));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _config = config ?? throw new ArgumentNullException(nameof(config));
+        }
+
+        /// <inheritdoc/>
+        public bool Supports(string resource) {
+            return _config.Query(resource, Provider).Any();
+        }
+
+        /// <inheritdoc/>
+        public async Task<TokenResultModel> GetTokenForAsync(string resource,
+            IEnumerable<string> scopes) {
+            if (_ctx.HttpContext?.User == null) {
+                return null;
+            }
+            var schemes = await _schemes.GetAllSchemesAsync();
+            if (!schemes.Any(s => s.Name == Provider)) {
+                return null;
+            }
+            foreach (var config in _config.Query(resource, Provider)) {
+                var decorator = CreateConfidentialClientApplication(_ctx.HttpContext.User,
+                    config, CreateRedirectUrl());
+                try {
+                    var result = await AcquireTokenSilentAsync(decorator.Client,
+                        _ctx.HttpContext.User, GetScopes(config, scopes), config.TenantId);
+                    if (result != null) {
+                        _logger.Debug(
+                            "Successfully acquired token {resource} with {config}.",
+                            resource, config.GetName());
+                        return result.ToTokenResult();
+                    }
+                }
+                catch (MsalUiRequiredException e) {
```

Comment:
> catch (MsalUiRequiredException e) { [](start = 15, length = 36)

is this catch covering all possible exceptions? or we want to propagate up other exceptions?

Reply:
Propagate all - this is a by design exception that lets you know to pull up ux.